### PR TITLE
Update to Mbedtls 2.16.0

### LINKED ISSUE #11: Update to MbedTLS 2.16.0
MbedTLS 2.16.0 was released recently: https://tls.mbed.org/tech-updates/releases/mbedtls-2.16.0-2.7.9-and-2.1.18-released

See https://github.com/fortanix/rust-mbedtls/#mbedtls-version-updates

### PR DESCRIPTION
**Pull request by jack-fortanix**
The first commit just replaces `mbedtls-sys/vendor` with the contents of 2.16.0 tarball. All changes by me are in the following commits.

Closes #11

### CODE REVIEW

File: mbedtls-sys/src/types.rs [hunk 1]
```diff
@@ -104,6 +104,9 @@ pub use self::libc::pthread_mutex_t;
 #[cfg(feature = "time")]
 pub use self::libc::time_t;
 
+#[cfg(feature = "time")]
+pub use self::libc::tm;
```
Comment (jethrogb) [inline mbedtls-sys/src/types.rs]: Use curly braces to condense into a single statement with the item above

File: mbedtls/src/error.rs [hunk 2]
```diff
@@ -113,7 +113,6 @@ error_enum!(
         AriaFeatureUnavailable = ERR_ARIA_FEATURE_UNAVAILABLE,
         AriaHwAccelFailed = ERR_ARIA_HW_ACCEL_FAILED,
         AriaInvalidInputLength = ERR_ARIA_INVALID_INPUT_LENGTH,
-        AriaInvalidKeyLength = ERR_ARIA_INVALID_KEY_LENGTH,
```
Comment (jethrogb) [inline mbedtls/src/error.rs]: I think you need to define `MBEDTLS_DEPRECATED_REMOVED`

File: mbedtls-sys/src/types.rs (hunk 1 above)
Comment (jethrogb) [inline mbedtls-sys/src/types.rs]: Use curly braces to condense into a single statement with the item above

**Comment by jethrogb** [comment]
You didn't follow the [update instructions](https://github.com/fortanix/rust-mbedtls#mbedtls-version-updates)

File: mbedtls/src/error.rs (hunk 2 above)
Comment (jack-fortanix) [reply mbedtls/src/error.rs]: It does not matter because even if the macro is defined, mbedtls won't return these errors anymore.

File: mbedtls-sys/build/config.rs [hunk 3]
```diff
@@ -36,24 +36,53 @@ pub const PREFIX: &'static str = r#"
...
+    ("MBEDTLS_CONFIG_H",                                  Defined),
     ("MBEDTLS_HAVE_ASM",                                  Defined),
-    ("MBEDTLS_NO_64BIT_MULTIPLICATION",                   Undefined),
     ("MBEDTLS_NO_UDBL_DIVISION",                          Undefined),
+    ("MBEDTLS_NO_64BIT_MULTIPLICATION",                   Undefined),
     ("MBEDTLS_HAVE_SSE2",                                 Undefined),
-    ("MBEDTLS_HAVE_TIME",                                 Undefined),
-    ("MBEDTLS_HAVE_TIME_DATE",                            Undefined),
+    ("MBEDTLS_HAVE_TIME",                                 Defined),
+    ("MBEDTLS_HAVE_TIME_DATE",                            Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: We definitely don't want this change

File: mbedtls-sys/build/config.rs [hunk 4]
```diff
@@ -139,7 +168,8 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
     ("MBEDTLS_ECP_DP_CURVE25519_ENABLED",                 Defined),
     ("MBEDTLS_ECP_DP_CURVE448_ENABLED",                   Defined),
     ("MBEDTLS_ECP_NIST_OPTIM",                            Defined),
-    ("MBEDTLS_ECDSA_DETERMINISTIC",                       Undefined),
+    ("MBEDTLS_ECP_RESTARTABLE",                           Undefined),
+    ("MBEDTLS_ECDSA_DETERMINISTIC",                       Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Nope

File: mbedtls-sys/build/config.rs [hunk 5]
```diff
@@ -154,10 +184,11 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
     ("MBEDTLS_PK_PARSE_EC_EXTENDED",                      Defined),
     ("MBEDTLS_ERROR_STRERROR_DUMMY",                      Defined),
     ("MBEDTLS_GENPRIME",                                  Defined),
-    ("MBEDTLS_FS_IO",                                     Undefined),
+    ("MBEDTLS_FS_IO",                                     Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this

File: mbedtls-sys/build/config.rs [hunk 6]
```diff
@@ -154,10 +184,11 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
...
     ("MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES",                Undefined),
-    ("MBEDTLS_NO_PLATFORM_ENTROPY",                       Defined),
+    ("MBEDTLS_NO_PLATFORM_ENTROPY",                       Undefined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this

File: mbedtls-sys/build/config.rs [hunk 7]
```diff
@@ -216,28 +248,28 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
...
+    ("MBEDTLS_DEBUG_C",                                   Defined),
     ("MBEDTLS_DES_C",                                     Defined),
     ("MBEDTLS_DHM_C",                                     Defined),
     ("MBEDTLS_ECDH_C",                                    Defined),
     ("MBEDTLS_ECDSA_C",                                   Defined),
-    ("MBEDTLS_ECJPAKE_C",                                 Defined),
+    ("MBEDTLS_ECJPAKE_C",                                 Undefined),
     ("MBEDTLS_ECP_C",                                     Defined),
-    ("MBEDTLS_ENTROPY_C",                                 Undefined),
+    ("MBEDTLS_ENTROPY_C",                                 Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this change.

File: mbedtls-sys/build/config.rs [hunk 8]
```diff
@@ -216,28 +248,28 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
...
     ("MBEDTLS_NIST_KW_C",                                 Undefined),
     ("MBEDTLS_MD_C",                                      Defined),
-    ("MBEDTLS_MD2_C",                                     Defined),
-    ("MBEDTLS_MD4_C",                                     Defined),
+    ("MBEDTLS_MD2_C",                                     Undefined),
+    ("MBEDTLS_MD4_C",                                     Undefined),
     ("MBEDTLS_MD5_C",                                     Defined),
     ("MBEDTLS_MEMORY_BUFFER_ALLOC_C",                     Undefined),
-    ("MBEDTLS_NET_C",                                     Undefined),
+    ("MBEDTLS_NET_C",                                     Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this change.

File: mbedtls-sys/build/config.rs [hunk 9]
```diff
@@ -248,7 +280,7 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
     ("MBEDTLS_PKCS5_C",                                   Defined),
     ("MBEDTLS_PKCS11_C",                                  Undefined),
     ("MBEDTLS_PKCS12_C",                                  Defined),
-    ("MBEDTLS_PLATFORM_C",                                Undefined),
+    ("MBEDTLS_PLATFORM_C",                                Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this change.

File: mbedtls-sys/build/config.rs [hunk 10]
```diff
@@ -262,7 +294,7 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
     ("MBEDTLS_SSL_SRV_C",                                 Defined),
     ("MBEDTLS_SSL_TLS_C",                                 Defined),
     ("MBEDTLS_THREADING_C",                               Undefined),
-    ("MBEDTLS_TIMING_C",                                  Undefined),
+    ("MBEDTLS_TIMING_C",                                  Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: Undo this change.

File: mbedtls-sys/build/config.rs [hunk 11]
```diff
@@ -290,33 +322,43 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
...
     ("MBEDTLS_SSL_CACHE_DEFAULT_MAX_ENTRIES",             Undefined), // default: 50
     ("MBEDTLS_SSL_MAX_CONTENT_LEN",                       Undefined), // default: 16384
     ("MBEDTLS_SSL_IN_CONTENT_LEN",                        Undefined), // default: 16384
     ("MBEDTLS_SSL_OUT_CONTENT_LEN",                       Undefined), // default: 16384
+    ("MBEDTLS_SSL_DTLS_MAX_BUFFERING",                    Undefined), // default: 32768
     ("MBEDTLS_SSL_DEFAULT_TICKET_LIFETIME",               Undefined), // default: 86400
     ("MBEDTLS_PSK_MAX_LEN",                               Undefined), // default: 32
     ("MBEDTLS_SSL_COOKIE_TIMEOUT",                        Undefined), // default: 60
-    ("MBEDTLS_SSL_CIPHERSUITES",                          Undefined), // no default
+    ("MBEDTLS_SSL_CIPHERSUITES",                          Undefined), // default: MBEDTLS_TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: “The value below is only an example, not the default.”

File: mbedtls-sys/build/config.rs (hunk 3 above)
Comment (jack-fortanix) [reply mbedtls-sys/build/config.rs]: If I understand correctly, you mean edit vendor config.h and then regenerate config.rs? Or just edit config.rs and leave the vendor version untouched?

It is not clear to me if config.rs is supposed to reflect the contents of config.h or to potentially override it.

File: mbedtls-sys/build/config.rs [hunk 12]
```diff
@@ -290,33 +322,43 @@ pub const DEFAULT_DEFINES: &'static [CDefine] = &[
...
     ("MBEDTLS_X509_MAX_INTERMEDIATE_CA",                  Undefined), // default: 8
     ("MBEDTLS_X509_MAX_FILE_PATH_LEN",                    Undefined), // default: 512
-    ("MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES",    Undefined),
+    ("MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES",    Defined),
```
Comment (jethrogb) [inline mbedtls-sys/build/config.rs]: I'm not sure about this change

Comment (jack-fortanix) [reply mbedtls-sys/build/config.rs]: If we're not supporting it now, we probably don't want to start supporting it.

Like I said, I assumed config.rs was supposed to exactly reflect config.h. If that's not the case, then we probably want to leave this turned off.

File: mbedtls/src/error.rs (hunk 2 above)
Comment (jethrogb) [reply mbedtls/src/error.rs]: Ah I see. If we've never released a version then we haven't change the public API. So it's ok to remove now. Thanks for the explanation.

File: mbedtls-sys/build/config.rs (hunk 3 above)
Comment (jethrogb) [reply mbedtls-sys/build/config.rs]: The `config.h` in `vendor/` is not used. The `config.h` is generated at build time from `config.rs`. All defines in `config.h` should be represented in `config.rs`, but the values may differ to suit our needs.

**Comment by jethrogb** [comment]
r=me after config.rs is fixed
